staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```diff
@@ -170,70 +171,78 @@ func (c *Controller) sync(name string) error {
 		reason = "update"
 	}
 	regenerationCounter.With(map[string]string{"crd": name, "reason": reason})
-	return c.updateSpecLocked()
+	return c.buildMergeSpecLocked()
 }
 
-func buildVersionSpecs(crd *apiextensionsv1.CustomResourceDefinition, oldSpecs map[string]*spec.Swagger) (map[string]*spec.Swagger, bool, error) {
-	newSpecs := map[string]*spec.Swagger{}
-	anyChanged := false
+func buildVersionSpecs(crd *apiextensionsv1.CustomResourceDefinition) (map[string]cached.Data[*spec.Swagger], error) {
+	newSpecs := map[string]cached.Data[*spec.Swagger]{}
 	for _, v := range crd.Spec.Versions {
 		if !v.Served {
 			continue
 		}
-		spec, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true})
-		// Defaults must be pruned here for CRDs to cleanly merge with the static
-		// spec that already has defaults pruned
-		spec.Definitions = handler.PruneDefaults(spec.Definitions)
-		if err != nil {
-			return nil, false, err
-		}
-		newSpecs[v.Name] = spec
-		if oldSpecs[v.Name] == nil || !reflect.DeepEqual(oldSpecs[v.Name], spec) {
-			anyChanged = true
-		}
+		newSpecs[v.Name] = cached.NewStaticSource(func() cached.Result[*spec.Swagger] {
+			s, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true})
+			// Defaults must be pruned here for CRDs to cleanly merge with the static
+			// spec that already has defaults pruned
+			if err != nil {
+				return cached.NewResultErr[*spec.Swagger](err)
+			}
+			s.Definitions = handler.PruneDefaults(s.Definitions)
+			return cached.NewResultOK(s, fmt.Sprintf("%d", crd.Generation))
+		})
 	}
-	if !anyChanged && len(oldSpecs) == len(newSpecs) {
-		return newSpecs, false, nil
-	}
-
-	return newSpecs, true, nil
+	return newSpecs, nil
 }
 
-// updateSpecLocked aggregates all OpenAPI specs and updates openAPIService.
-// It is not thread-safe. The caller is responsible to hold proper lock (Controller.lock).
-func (c *Controller) updateSpecLocked() error {
-	crdSpecs := []*spec.Swagger{}
-	for _, versionSpecs := range c.crdSpecs {
-		for _, s := range versionSpecs {
-			crdSpecs = append(crdSpecs, s)
+func (c *Controller) buildMergeSpecLocked() error {
+	crdSpecs := map[string]cached.Data[*spec.Swagger]{}
+	for g := range c.crdSpecs {
+		for v := range c.crdSpecs[g] {
+			crdSpecs[g+"/"+v] = c.crdSpecs[g][v]
 		}
 	}
-	mergedSpec, err := builder.MergeSpecs(c.staticSpec, crdSpecs...)
-	if err != nil {
-		return fmt.Errorf("failed to merge specs: %v", err)
-	}
-	return c.openAPIService.UpdateSpec(mergedSpec)
+
+	cache := cached.NewMerger(func(results map[string]cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		localCRDSpec := []*spec.Swagger{}
+		for k := range results {
+			localCRDSpec = append(localCRDSpec, results[k].Data)
+		}
+		mergedSpec, err := builder.MergeSpecs(c.staticSpec, localCRDSpec...)
+		if err != nil {
+			return cached.NewResultErr[*spec.Swagger](fmt.Errorf("failed to merge specs: %v", err))
+		}
+		return cached.NewResultOK(mergedSpec, uuid.New().String())
```

Comment:
Technically we could compute the etag as a hash of the structs (a bit expensive since inputs are not pre-marshalled like in the aggregator), but we only ever call buildMergeSpecLocked if a CRD has changed (generation counter incremented), so I think a uuid/time based etag should be okay.

Reply:
Add a comment in the code though.